---
title: "OVEX Data Analysis"
author: "Alex Handzel"
date: "1/12/2022"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

Libraries Imported

```{r}
library(readr)
library(ggplot2)
theme_set(theme_bw())
library(car)
library(dplyr)
library(tidyr)
library(data.table)
library(vegan)
library(tidyverse)
library(ggpubr)
library(pairwiseAdonis)
library(usedist)
#library(phangorn)
library(multcomp)
library(randomForest)
library(mcp)
library(emmeans)
library(nlme)
library(caret)
library(umap)
library(randomForestExplainer)
library(readxl)
library(rstatix)
library(svglite)
```

Alpha Diversity Analysis: Faiths PD

```{r}
#Importing Alpha Diversity Data from Qiime2
alpha <- read_csv("ovexalphadiversity.csv")



#Unable to normalize data, based on the case of large sample size (n>30), can ignore the distribution of the data and use parametric tests. 
#http://www.sthda.com/english/wiki/normality-test-in-r#case-of-large-sample-sizes

ggqqplot(alpha, 'faith_pd')

#Normality and Heteroscedasticity check: GOOD
faith_heter_check <- lm(faith_pd ~ Treatment*Time, data = alpha)
plot(faith_heter_check)


Week = as.factor(alpha$Time)
faith <- ggplot(alpha, aes(x=Treatment, y=faith_pd, fill = Week)) + 
  geom_boxplot()+
  theme_classic()+
  ylab("Faith's Phylogenetic Diversity")+
  ggtitle("Faith's PD by Treatment")

#faith

ovxdht_faith=alpha %>%
  filter(Treatment == 'OVX + DHT')
shamdht_faith=alpha %>%
  filter(Treatment == 'SHAM + DHT')
ovxp_faith=alpha %>%
  filter(Treatment == 'OVX + P')
shamp_faith=alpha %>%
  filter(Treatment == 'SHAM + P')

###OVX + DHT
#Scatter plot (R2 and p-value)
scatt_faith_ovxdht= ggplot(ovxdht_faith, aes(x=Time, y=faith_pd))+
  geom_point()+
  #theme_classic()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 17)+
  ylab("Faith's PD")+
  xlab("Week")+
  ylim(4, 17)+
  ggtitle("OVX+DHT")
#scatt_faith_ovxdht

scatt_even_ovxdht = ggplot(ovxdht_faith, aes(x=Time, y=pielou_evenness))+
  geom_point()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 0.99)+
  ylab("Evenness")+
  xlab("Week")+
  ylim(0.92, 0.99)+
  ggtitle("OVX+DHT")
#scatt_even_ovxdht
#ggsave(filename = "scatt_faith_OVXDHT.svg", plot = scatt_faith_ovxdht)


###SHAM + DHT
scatt_faith_shamdht= ggplot(shamdht_faith, aes(x=Time, y=faith_pd))+
  geom_point()+
  #theme_classic()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 17)+
  ylab("Faith's PD")+
  xlab("Week")+
  ylim(4, 17)+
  ggtitle("SHAM+DHT")
#scatt_faith_shamdht
#ggsave(filename = "scatt_faith_SHAMDHT.svg", plot = scatt_faith_shamdht)

scatt_even_shamdht = ggplot(shamdht_faith, aes(x=Time, y=pielou_evenness))+
  geom_point()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 0.99)+
  ylab("Evenness")+
  xlab("Week")+
  ylim(0.92, 0.99)+
  ggtitle("SHAM+DHT")
#scatt_even_shamdht


### OVX + P
scatt_faith_ovxp= ggplot(ovxp_faith, aes(x=Time, y=faith_pd))+
  geom_point()+
  #theme_classic()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 17)+
  ylab("Faith's PD")+
  xlab("Week")+
  ylim(4, 17)+
  ggtitle("OVX+P")
#scatt_faith_ovxp

scatt_even_ovxp = ggplot(ovxp_faith, aes(x=Time, y=pielou_evenness))+
  geom_point()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 0.99)+
  ylab("Evenness")+
  xlab("Week")+
  ylim(0.92, 0.99)+
  ggtitle("OVX+P")
#scatt_even_ovxp
#ggsave(filename = "scatt_faith_OVXP.svg", plot = scatt_faith_ovxp)


### SHAM + P
scatt_faith_shamp= ggplot(shamp_faith, aes(x=Time, y=faith_pd))+
  geom_point()+
  #theme_classic()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 17)+
  ylab("Faith's PD")+
  xlab("Week")+
  ylim(4, 17)+
  ggtitle("SHAM+P")
#scatt_faith_shamp

scatt_even_shamp = ggplot(shamp_faith, aes(x=Time, y=pielou_evenness))+
  geom_point()+
  geom_smooth(method=lm, se=TRUE)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 0.99)+
  ylab("Evenness")+
  xlab("Week")+
  ylim(0.92, 0.99)+
  ggtitle("SHAM+P")
#scatt_even_shamp
#ggsave(filename = "scatt_faith_SHAMP.svg", plot = scatt_faith_shamp)

alpha_figure = ggpubr::ggarrange(scatt_faith_ovxdht,scatt_faith_ovxp,scatt_faith_shamdht,scatt_faith_shamp, ncol = 2,nrow=2,labels = c("A", "B", "C", "D"))
#alpha_figure

alpha_figure_even = ggpubr::ggarrange(scatt_even_ovxdht,scatt_even_ovxp,scatt_even_shamdht,scatt_even_shamp, ncol = 2,nrow=2,labels = c("A", "B", "C", "D"))
#alpha_figure_even
#ggsave(filename = "scatter_even.svg", plot = alpha_figure_even)

#Linear Mixed model with mice as a random effect
alt_model <- lme(faith_pd ~ Treatment*Time, random = ~ 1 | Actual.Contents, data = alpha)
alt_model1 <- lme(pielou_evenness ~ Treatment*Time, random = ~ 1 | Actual.Contents, data = alpha)
anova(alt_model1)


```

Alpha Diversity Analysis: Pielou's Evenness

```{r}

#Unable to normalize data, based on the case of large sample size (n>30), can ignore the distribution of the data and use parametric tests. 
#http://www.sthda.com/english/wiki/normality-test-in-r#case-of-large-sample-sizes

ggqqplot(alpha, 'pielou_evenness')

#Normality check and heteroscedasticity check: Good
even_heter_check <- lm(pielou_evenness ~ Treatment*Time, data = alpha)
plot(even_heter_check)


#BoxPlot Evenness by treatment type with time as a color
even <- ggplot(alpha, aes(x=Treatment, y=pielou_evenness, fill = Week)) + 
  geom_boxplot()+
  theme_classic()+
  ylab("Evenness")+
  ggtitle("Evenness by Treatment")

even

#QQplot to visualize normality among the groups
ggqqplot(alpha, 'pielou_evenness', facet.by = 'Treatment')

#Scatter plot (R2 and p-value)
scatt_even = ggplot(alpha, aes(x=Time, y=pielou_evenness))+
  geom_point()+
  theme_classic()+
  geom_smooth(method=lm, se=TRUE)+
  facet_grid(~Treatment)+
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),label.y = 0.99)+
  ylab("Evenness")+
  xlab("Week")
  
#ggsave(filename = "scatt_even.svg", plot = scatt_even)
#scatt_even



###Will have to separate out by treatment to get P-value for figure above
#Linear Model checking for interactions between treatment, time = No interactions that are significant use type II
lmm_even = lme(pielou_evenness ~ Time*Treatment, random = ~1| Actual.Contents, data = alpha)
summary(lmm_even)
anova(lmm_even)

```

Beta Diversity Analysis:

Imported Files/Data Cleaning

```{r}
#Importing Taxa table
taxa = read.table('taxa.tsv', sep = '\t', header = T, check.names = F)

#Importing OTU table
OTUtableA <- read.csv('./biomtxt/table.from_biom.tsv',
                     sep = '\t',header = T, row.names = 1,
                     check.names = F)

#Checking for zero values in OTU table
tftabA<-OTUtableA==0
tftabA[,1]
tftabA[tftabA] <- 1
tftabA[tftabA==F] <- 0

#Removing taxa that have a sum less than 246.65 based on rarifaction curve. 
CUTOFF <- 246.65
sum(rowSums(tftabA)>CUTOFF)
sum(rowSums(tftabA)<CUTOFF)
zerofiltOTUa <- OTUtableA[rowSums(tftabA)<CUTOFF,]


#Importing Sequencing Key
sequencekey <- read.csv('sequencingkey1.tsv',
                     sep = '\t',header = T)

#Sequencing Key data clean up:
#1) Sample ID changed from 56 to X56
#2) Removed columns that add no value to data analysis
#3) Convert Treatment and Time to factors
sequencekey$SampleID = sub("^", "X", sequencekey$SampleID)

sequencekey = subset(sequencekey, select = -c(BarcodeSequence, LinkerPrimerSequence, Plate, Well, Primer, Cage))
sequencekey$Treatment = as.factor(sequencekey$Treatment)
sequencekey$Time = as.factor(sequencekey$Time)
sequencekey$Actual.Contents = as.factor(sequencekey$Actual.Contents)

#Removing bad samples and controls from the sequencing key
new_sequecingkey=sequencekey %>%
  filter(!SampleID == 'X56', !SampleID == 'X57',!SampleID == 'X58',!SampleID == 'X249',!SampleID == 'X250',!SampleID == 'X251', !SampleID == 'X252',!SampleID == 'X253', !SampleID == 'X20', !SampleID == 'X24', !SampleID == 'X51',!SampleID == 'X74', !SampleID == 'X149', !SampleID == 'X179', !SampleID == 'X200', !SampleID == 'X242')

```

Zero Replacement and CLR Transformation

```{r}
#Zero Replacement
zerofiltOTUa[zerofiltOTUa == 0] = 0.05

#CLR Transformation Function
clr = function(x) sweep(log(x), 1, rowMeans(log(x)), "-")
zerofiltOTUa
#Transpose dataframe, perform the CLR transformation, transpose back
bacteria_tx = data.frame(t(clr(t(zerofiltOTUa))))

#Removing control Samples/bad samples(Large Percentage of Unknown bacteria)
#Cannot remove bad/control samples until after CLR transformation: will change the row means and affect the data
bacteria_tx = subset(bacteria_tx, select = -c(`X56`, `X57`, `X58`, `X249`, `X250`, `X251`, `X252`, `X253`, `X20`, `X24`, `X51`, `X74`, `X149`, `X179`, `X200`, `X242`))


bacteria_tx
#bacteria_new = tibble::rownames_to_column(bacteria_tx, "OTUID")

```

Random Forest Feature Selection

```{r}
bacteria_new = data.frame(t(bacteria_tx))

#list of time zero sample ID
time_zero_names = c("X1", "X2", "X3", "X11", "X12", "X13", "X14", "X15", "X16", "X17",
                    "X18", "X19", "X21", "X22", "X23", "X25", "X26","X27", "X28","X29","X30",
                    "X31","X32","X33","X34","X35","X36","X37","X38","X39","X40",
                    "X41","X42","X43","X44","X45","X46","X257")

#Removing all time zero sample ID for RF
bact_rf = bacteria_new[!(row.names(bacteria_new) %in% time_zero_names),]

new_sequecingkey1 = data.frame(new_sequecingkey, row.names = 1)
rf.data = merge(bact_rf, new_sequecingkey1, by=0)
rf.data$Time = droplevels(rf.data$Time)
rf.data$Treatment = droplevels(rf.data$Treatment)
rf.data = data.frame(rf.data, row.names = 1)
rf.data = subset(rf.data, select = -c(`Time`, `Actual.Contents`))
rf.data

#RandomForest
set.seed(42)
erie.classify <- randomForest(Treatment ~ ., data=rf.data, importance=TRUE, scale=FALSE)
erie.classify

#Creating dataframe of importance 
impo = data.frame(importance(erie.classify))

varImpPlot(erie.classify, n.var = 100)

#Sorting by decreasing mean gini
impo = impo[order((impo$MeanDecreaseGini),decreasing=TRUE),]

#Selecting top 50 Features
impo.50 = impo[1:50, ]

#Feature Name to column to merge taxa table
impo.50 = tibble::rownames_to_column(impo.50, "feature")
impo.50$feature = gsub("X", "", impo.50$feature)
impo.50 = merge(impo.50, taxa, by="feature")
impo.50 = impo.50 %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")
impo.50

#impo.50 %>% write_csv(paste(lubridate::today(), 'rf_importance.csv', sep = "_"))
rf_class <- read.csv("C:/Users/Alex/Documents/Thesis/Tables(results)/2022-09-21_rf_importance.csv")
rf_class

#Creating barplot to show descending importance based on OTU
rf_img = ggplot(data = rf_class, aes(x = reorder(Classification, MeanDecreaseGini), y=MeanDecreaseGini))+
  geom_bar(stat = "identity", fill='steelblue', width = 0.7)+
  coord_flip()+
  ylab("Importance")+
  xlab("OTUs")+
  ylim(0,0.71)

rf_img
#ggsave(filename = "rf_img.svg", plot = rf_img)

#Selecting top n names for use in differential taxa analysis
names_rf = impo %>% rownames_to_column() %>% top_n(50, MeanDecreaseGini) %>% pull(rowname)
#names_rf





```

Line Plots of 6 different classifications

```{r}
line_plots = data.frame(t(bacteria_tx))

#Selecting only Schaedleri species that are statistically significant
line_plots = subset(line_plots, select = c(X8a308a269f4a15e0885ddfba3d4c408e, X8a9a536075ca4d38f077c4f22d915770, X9d607ec0281187001a9d6e5e22f9561d
, X2e59f1b99fab85cb63bce2f35a5ca838, X64547667c88735701e168289a2a66718, X54df10b2c671e1384b096ab492a3c85b,
X896ff78f6a5e46e65321c4b3c8e0635d, f39682f9eb7628209f54c4fc7276617c, X61f41b7546d47c5a7495d40062c6b1e5,f4496608bb586eb4cfd0969841385e32, ace220b7598137d9dbab272c9644ef45
))

#Merging clr counts with sequencing key
Seq_key_line = data.frame(new_sequecingkey, row.names = 1)
line_plots_w_key = merge(line_plots, new_sequecingkey1, by=0)
line_plots_w_key = data.frame(line_plots_w_key, row.names = 1)

#Calculating mean for each group at each time point
mean_var = line_plots_w_key %>%
  group_by(Time, Treatment)%>%
  summarise(across(everything(), mean))
mean_var

#Plotting mean vs time point for each Schaedleri Species


s24_7_7 = ggplot(data = mean_var, aes(x=Time, y=X8a9a536075ca4d38f077c4f22d915770, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("f__S24-7_7")

s24_7_7
  

Lachno_1 = ggplot(data = mean_var, aes(x=Time, y=X9d607ec0281187001a9d6e5e22f9561d, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("f__Lachnospiraceae_1")

Lachno_1


schaed_3 = ggplot(data = mean_var, aes(x=Time, y=X8a308a269f4a15e0885ddfba3d4c408e, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("s__Schaedleri_3")

schaed_3 


allobaculum = ggplot(data = mean_var, aes(x=Time, y=X2e59f1b99fab85cb63bce2f35a5ca838, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("g__Allobaculum")

allobaculum


lactobac_2 = ggplot(data = mean_var, aes(x=Time, y=X64547667c88735701e168289a2a66718, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("g__Lactobacillus_2")

lactobac_2 

riken_2 = ggplot(data = mean_var, aes(x=Time, y=X54df10b2c671e1384b096ab492a3c85b, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("f__Rikenellaceae_2")


riken_2

avg_clr = ggpubr::ggarrange(s24_7_7, Lachno_1, schaed_3,allobaculum, 
lactobac_2, riken_2, ncol = 2,nrow=3,labels = c("A", "B", "C", "D", "E", "F"))
avg_clr
#ggsave(filename = "avg_clr_abu.svg", plot = avg_clr)

```

Line plots after taxa analysis 
```{r}

##Good for OVX+P
lacto_2_top = ggplot(data = mean_var, aes(x=Time, y=f39682f9eb7628209f54c4fc7276617c
, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("Lactobacillus")

lacto_2_top


##Good for SHAM+DHT
Riken_another = ggplot(data = mean_var, aes(x=Time, y=ace220b7598137d9dbab272c9644ef45
, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("Rikenelleceae")

Riken_another

##Good for OVTDHT group
s247_f44 = ggplot(data = mean_var, aes(x=Time, y=f4496608bb586eb4cfd0969841385e32
, group=Treatment))+
  geom_line(aes(color=Treatment))+
  geom_point(aes(color=Treatment))+
  ylim(-1,7.5)+
  ylab("Avg CLR Abundance")+
  xlab("Time (Week)")+
  ggtitle("S24-7")

s247_f44

empty1 = ggplot()
empty1

empty2 = ggplot()
empty2

empty3 = ggplot()
empty3


avg_clr_after_taxa_analysis = ggpubr::ggarrange(lacto_2_top, empty1, Riken_another, empty2, s247_f44, empty3, ncol = 2,nrow=3,labels = c("A", "", "B", "", "C", ""))
#avg_clr_after_taxa_analysis
#ggsave(filename = "new_Other_3_clr_abund.svg", plot = avg_clr_after_taxa_analysis)


```



NMDS Analysis Using Euclidean Distance:

```{r}
set.seed(42)
#NMDS
bact_m = as.matrix(t(bacteria_tx))
Bact_nmds = metaMDS(bact_m, distance = "euclidean")
Bact_nmds
stressplot(Bact_nmds)
#bact_fit = envfit(Bact_nmds, bact_m)
#head(bact_fit)
bact_scores = as.data.frame(scores(Bact_nmds, display = 'sites'))
bact_scores

#Merging Sequencing key to nmds scores. 
bact_scores = cbind(SampleID = rownames(bact_scores), bact_scores)
rownames(bact_scores)<-1:nrow(bact_scores)

merged_nmds = merge(bact_scores, sequencekey, by="SampleID")
merged_nmds$Time = as.factor(merged_nmds$Time)

merged_nmds

```

Plotting NMDS Without time zero

```{r}

#Plotting without time zero, color based on treatment 
not0 = subset(merged_nmds, Time !=0)
head(not0)

T_no0 = ggplot(data = not0)+
  theme_bw()+
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+
  ylim(-17, 17)+
  xlim(-17, 17)+
  ggtitle("NMDS plot Excluding Time Zero")+
  theme(plot.title = element_text(hjust = 0.5))

T_no0

```

Plotting NMDS and separating plots based on time points

```{r}

#Separating data based on Time points
t0 = subset(merged_nmds, Time == 0)

mod = lm(Time ~ NMDS1 + NMDS2, data = t0)
t1 = subset(merged_nmds, Time == 1)
t2 = subset(merged_nmds, Time == 2)
t3 = subset(merged_nmds, Time == 3)
t4 = subset(merged_nmds, Time == 4)
t5 = subset(merged_nmds, Time == 5)


#Plotting Individual Time points and colored by Treatment
T_0 = ggplot(data = t0) +
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment, size = 1))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+  
  xlim(-125, 75)+
  ylim(-70, 60)+
  ggtitle("Week 0")+
  theme(plot.title = element_text(hjust = 0.5))
T_0
T_0 %>%
  ggexport(filename = "nmds_t0.pdf")

T_1 = ggplot(data = t1) +
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment, size = 1))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+
  xlim(-125, 75)+
  ylim(-70, 60)+
  ggtitle("Week 1")+
  theme(plot.title = element_text(hjust = 0.5))
T_1
T_1 %>%
  ggexport(filename = "nmds_t1.pdf")

T_2 = ggplot(data = t2) +
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment, size = 1))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+
  xlim(-125, 75)+
  ylim(-70, 60)+
  ggtitle("Week 2")+
  theme(plot.title = element_text(hjust = 0.5))
T_2
T_2 %>%
  ggexport(filename = "nmds_t2.pdf")

T_3 = ggplot(data = t3) +
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment, size = 1))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+
  xlim(-125, 75)+
  ylim(-70, 60)+
  ggtitle("Week 3")+
  theme(plot.title = element_text(hjust = 0.5))
T_3
T_3 %>%
  ggexport(filename = "nmds_t3.pdf")

T_4 = ggplot(data = t4) +
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment, size = 1))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+
  xlim(-125, 75)+
  ylim(-70, 60)+
  ggtitle("Week 4")+
  theme(plot.title = element_text(hjust = 0.5))
T_4
T_4 %>%
  ggexport(filename = "nmds_t4.pdf")

T_5 = ggplot(data = t5) +
  geom_point(aes(x = NMDS1, y = NMDS2, color = Treatment, size=1))+
  #stat_ellipse(aes(x = NMDS1, y = NMDS2,color = Treatment))+
  xlim(-125, 75)+
  ylim(-70, 60)+
  ggtitle("Week 5")+
  theme(plot.title = element_text(hjust = 0.5))
T_5 %>%
  ggexport(filename = "nmds_t5.pdf")

NMDS_figure = ggpubr::ggarrange(T_0,T_1,T_2,T_3,T_4,T_5, ncol = 2,nrow=3,labels = c("A", "B", "C", "D", "E", "F"))

```

Permanova on all Euclidean Distances

```{r}
set.seed(42)
#Permanova:
bact_dist = vegdist(bact_m, method = 'euclidian')


bact_perm_treat = adonis2(bact_dist~Treatment*Time, data = new_sequecingkey, permutations = 999, method = 'euclidian')
bact_perm_treat

bact_dist = vegdist(bact_m, method = 'euclidian')
bact_perm_time = adonis2(bact_dist~Treatment, data = new_sequecingkey, permutations = 999, method = 'euclidian')
bact_perm_time

#Pairwise Permanova
pairwise.adonis2(bact_dist~Treatment, data = new_sequecingkey, permutations = 999, method = 'euclidian')

```

Week 0 Analysis: 1) Permanova 2) Ad Hoc Pairwise Analysis 3)
Beta-Dispersion

```{r}
#install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
#time = 0 Permanova
bacteria_time0 = subset(bacteria_tx, select = c(`X1`, `X2`, `X3`, `X11`, `X12`, `X13`, `X14`, `X15`, `X16`, `X17`, 
                                             `X18`, `X19`, `X21`, `X22`, `X23`, `X25`, `X26`,`X27`,`X28`,`X29`,`X30`,
                                             `X31`,`X32`,`X33`,`X34`,`X35`,`X36`,`X37`,`X38`,`X39`,`X40`,
                                             `X41`,`X42`,`X43`,`X44`,`X45`,`X46`,`X257`))
sequencingkey_time0=new_sequecingkey %>%
  filter(Time == 0)
sequencingkey_time0
#head(sequencingkey_time0)

#transposing and making it a matrix
bact_m_time0 = as.matrix(t(bacteria_time0))
bact_dist_time0 = vegdist(bact_m_time0, method = 'euclidian')

set.seed(42)
#1) Permanova
bact_perm_time0=adonis2(bact_dist_time0~Treatment, data = sequencingkey_time0, permutations = 9999, method = 'euclidian')
bact_perm_time0
bact_perm_time0$R2[1]
bact_perm_time0$`Pr(>F)`[1]

#2) Ad Hoc Pairwise Analysis
pairwise_bac_time0 = pairwise.adonis2(bact_dist_time0~Treatment, data = sequencingkey_time0, method = 'euclidian')
pairwise_bac_time0

#3) Beta-dispersion: Distance from centroid of Group
#Calculating Beta-dispersion Distance from centroiod of Group)
bact_betadist_treat0 = betadisper(bact_dist_time0, group = sequencingkey_time0$Treatment)
distances_microbe_treat0 = data.frame(bact_betadist_treat0[['distances']])
distances_microbe_treat0 = cbind(distances_microbe_treat0, group = sequencingkey_time0$Treatment)
colnames(distances_microbe_treat0) = c('Distance_to_Centroid', 'Treatment')
#head(distances_microbe_treat0)

#Permutation test for homogeneity of Multivariate dispesions
test_bact_betadist_treatment0 = permutest(bact_betadist_treat0)
test_bact_betadist_treatment0

#Graphing Beta dispersions
B_Treat0 <- ggboxplot(data = distances_microbe_treat0, x = "Treatment", y = "Distance_to_Centroid" ) +
theme_classic() +
geom_point()
B_Treat0

```

Week 1 Analysis: 1) Permanova 2) Ad Hoc Pairwise Analysis 3)
Beta-Dispersion

```{r}
#time = 1 Permanova
bacteria_time1 = subset(bacteria_tx, select = c(`X4`, `X5`, `X6`, `X47`, `X48`, `X49`, `X50`, `X52`, `X53`, 
                                             `X54`, `X55`, `X59`, `X60`, `X61`, `X62`, `X63`,`X64`,`X65`,`X66`,
                                             `X67`,`X68`,`X69`,`X70`,`X71`,`X72`,`X73`,`X75`,`X76`,`X77`,
                                             `X78`,`X79`,`X80`,`X81`,`X82`,`X83`,`X84`,`X85`,`X85`))
#dim(bacteria_time1)
sequencingkey_time1=new_sequecingkey %>%
  filter(Time == 1)


set.seed(42)
#transposing and making it a matrix
bact_m_time1 = as.matrix(t(bacteria_time1))
bact_dist_time1 = vegdist(bact_m_time1, method = 'euclidian')

#1) Permanova
bact_perm_time1=adonis2(bact_dist_time1~Treatment, data = sequencingkey_time1, permutations = 9999, method = 'euclidian')
bact_perm_time1
bact_perm_time1$R2[1]
bact_perm_time1$`Pr(>F)`[1]

#2) Ad Hoc Pairwise Analysis
pairwise_bac_time1 = pairwise.adonis2(bact_dist_time1~Treatment, data = sequencingkey_time1, permutations = 9999, method = 'euclidian')
pairwise_bac_time1
#3) Beta-dispersion: Distance from centroid of Group
#Calculating Beta-dispersion Distance from centroiod of Group)
bact_betadist_treat1 = betadisper(bact_dist_time1, group = sequencingkey_time1$Treatment)
distances_microbe_treat1 = data.frame(bact_betadist_treat1[['distances']])
distances_microbe_treat1 = cbind(distances_microbe_treat1, group = sequencingkey_time1$Treatment)
colnames(distances_microbe_treat1) = c('Distance_to_Centroid', 'Treatment')
#head(distances_microbe_treat1)

#Permutation test for homogeneity of Multivariate dispesions
test_bact_betadist_treatment1 = permutest(bact_betadist_treat1)
test_bact_betadist_treatment1

#Graphing Beta dispersions
B_Treat1 <- ggboxplot(data = distances_microbe_treat1, x = "Treatment", y = "Distance_to_Centroid" ) +
theme_classic() +
geom_point()
B_Treat1

```

Week 2 Analysis: 1) Permanova 2) Ad Hoc Pairwise Analysis 3)
Beta-Dispersion

```{r}


bacteria_time2 = subset(bacteria_tx, select = c(`X7`, `X8`, `X87`, `X88`, `X89`, `X90`, `X91`, `X92`, `X93`, 
                                             `X94`, `X95`, `X96`, `X97`, `X99`, `X100`, `X101`,`X102`,`X103`,`X104`,
                                             `X105`,`X106`,`X107`,`X108`,`X109`,`X110`,`X111`,`X112`,`X113`,`X114`,
                                             `X115`,`X116`,`X117`,`X118`,`X119`,`X120`,`X121`,`X122`,`X125`, `X248`, `X255`))

sequencingkey_time2=new_sequecingkey %>%
  filter(Time == 2)

set.seed(42)
#transposing and making it a matrix
bact_m_time2 = as.matrix(t(bacteria_time2))
bact_dist_time2 = vegdist(bact_m_time2, method = 'euclidian')

#1) Permanova
bact_perm_time2=adonis2(bact_dist_time2~Treatment, data = sequencingkey_time2, permutations = 9999, method = 'euclidian')
bact_perm_time2

#2) Ad Hoc Pairwise Analysis
pairwise_bac_time2 = pairwise.adonis2(bact_dist_time2~Treatment, data = sequencingkey_time2, permutations = 9999, method = 'euclidian')
pairwise_bac_time2

#3) Beta-dispersion: Distance from centroid of Group
#Calculating Beta-dispersion Distance from centroid of Group)
bact_betadist_treat2 = betadisper(bact_dist_time2, group = sequencingkey_time2$Treatment)
distances_microbe_treat2 = data.frame(bact_betadist_treat2[['distances']])
distances_microbe_treat2 = cbind(distances_microbe_treat2, group = sequencingkey_time2$Treatment)
colnames(distances_microbe_treat2) = c('Distance_to_Centroid', 'Treatment')
#head(distances_microbe_treat2)

#Permutation test for homogeneity of Multivariate dispersion
test_bact_betadist_treatment2 = permutest(bact_betadist_treat2)
test_bact_betadist_treatment2

#Graphing Beta dispersions
B_Treat2 <- ggboxplot(data = distances_microbe_treat2, x = "Treatment", y = "Distance_to_Centroid" ) +
theme_classic() +
geom_point()
B_Treat2

```

Week 3 Analysis: 1) Permanova 2) Ad Hoc Pairwise Analysis 3)
Beta-Dispersion

```{r}
bacteria_time3 = subset(bacteria_tx, select = c(`X126`, `X127`, `X128`, `X129`, `X130`, `X131`, `X132`, `X133`, `X134`, 
                                             `X135`, `X136`, `X137`, `X138`, `X139`, `X140`, `X141`,`X142`,`X143`,`X144`,
                                             `X145`,`X147`,`X148`,`X150`,`X151`,`X152`,`X153`,`X154`,`X155`,
                                             `X156`,`X157`,`X158`,`X159`,`X160`,`X161`,`X162`,`X163`,`X164`, `X165`, `X166`))

sequencingkey_time3=new_sequecingkey %>%
  filter(Time == 3)

set.seed(42)
#transposing and making it a matrix
bact_m_time3 = as.matrix(t(bacteria_time3))
bact_dist_time3 = vegdist(bact_m_time3, method = 'euclidian')

#1) Permanova
bact_perm_time3=adonis2(bact_dist_time3~Treatment, data = sequencingkey_time3, permutations = 9999, method = 'euclidian')
bact_perm_time3

#2) Ad Hoc Pairwise Analysis
pairwise_bac_time3 = pairwise.adonis2(bact_dist_time3~Treatment, data = sequencingkey_time3, permutations = 9999, method = 'euclidian')
pairwise_bac_time3

#3) Beta-dispersion: Distance from centroid of Group
#Calculating Beta-dispersion Distance from centroiod of Group)
bact_betadist_treat3 = betadisper(bact_dist_time3, group = sequencingkey_time3$Treatment)
distances_microbe_treat3 = data.frame(bact_betadist_treat3[['distances']])
distances_microbe_treat3 = cbind(distances_microbe_treat3, group = sequencingkey_time3$Treatment)
colnames(distances_microbe_treat3) = c('Distance_to_Centroid', 'Treatment')
#head(distances_microbe_treat3)

#Permutation test for homogeneity of Multivariate dispesions
test_bact_betadist_treatment3 = permutest(bact_betadist_treat3)
test_bact_betadist_treatment3

#Graphing Beta dispersions
B_Treat3 <- ggboxplot(data = distances_microbe_treat3, x = "Treatment", y = "Distance_to_Centroid" ) +
theme_classic() +
geom_point()
B_Treat3


```

Week 4 Analysis: 1) Permanova 2) Ad Hoc Pairwise Analysis 3)
Beta-Dispersion

```{r}
bacteria_time4 = subset(bacteria_tx, select = c(`X167`, `X168`, `X169`, `X170`, `X171`, `X172`, `X173`, `X174`, `X175`, 
                                             `X176`, `X177`, `X178`, `X180`, `X181`, `X182`, `X183`,`X184`,`X185`,`X186`,
                                             `X187`,`X188`,`X189`,`X190`,`X191`,`X192`,`X193`,`X195`,`X196`,
                                             `X197`,`X198`,`X199`,`X201`,`X202`,`X203`,`X204`,`X205`,`X206`, `X207`))

sequencingkey_time4=new_sequecingkey %>%
  filter(Time == 4)
set.seed(42)
#transposing and making it a matrix
bact_m_time4 = as.matrix(t(bacteria_time4))
bact_dist_time4 = vegdist(bact_m_time4, method = 'euclidian')

#1) Permanova
bact_perm_time4=adonis2(bact_dist_time4~Treatment, data = sequencingkey_time4, permutations = 9999, method = 'euclidian')
bact_perm_time4

#2) Ad Hoc Pairwise Analysis
pairwise_bac_time4 = pairwise.adonis2(bact_dist_time4~Treatment, data = sequencingkey_time4, permutations = 9999, method = 'euclidian')
pairwise_bac_time4

#3) Beta-dispersion: Distance from centroid of Group
#Calculating Beta-dispersion Distance from centroiod of Group)
bact_betadist_treat4 = betadisper(bact_dist_time4, group = sequencingkey_time4$Treatment)
distances_microbe_treat4 = data.frame(bact_betadist_treat4[['distances']])
distances_microbe_treat4 = cbind(distances_microbe_treat4, group = sequencingkey_time4$Treatment)
colnames(distances_microbe_treat4) = c('Distance_to_Centroid', 'Treatment')
#head(distances_microbe_treat4)

#Permutation test for homogeneity of Multivariate dispesions
test_bact_betadist_treatment4 = permutest(bact_betadist_treat4)
test_bact_betadist_treatment4

#Graphing Beta dispersions
B_Treat4 <- ggboxplot(data = distances_microbe_treat4, x = "Treatment", y = "Distance_to_Centroid" ) +
theme_classic() +
geom_point()
B_Treat4

```

Week 5 Analysis: 1) Permanova 2) Ad Hoc Pairwise Analysis 3)
Beta-Dispersion

```{r}
bacteria_time5 = subset(bacteria_tx, select = c(`X208`, `X209`, `X210`, `X211`, `X212`, `X213`, `X214`, `X215`, `X216`, 
                                             `X217`, `X218`, `X219`, `X220`, `X221`, `X222`, `X223`,`X224`,`X225`,`X226`,
                                             `X227`,`X228`,`X229`,`X230`,`X231`,`X232`,`X233`,`X234`,`X235`,
                                             `X236`,`X237`,`X238`,`X239`,`X240`,`X241`,`X247`,`X243`,`X244`, `X245`, `X246`))

bacteria_time5
#Sequencing key containing only time = 5
sequencingkey_time5=new_sequecingkey %>%
  filter(Time == 5)

set.seed(42)
#transposing and making it a matrix
bact_m_time5 = as.matrix(t(bacteria_time5))
bact_dist_time5 = vegdist(bact_m_time5, method = 'euclidian')

#1) Permanova 
bact_perm_time5=adonis2(bact_dist_time5~Treatment, data = sequencingkey_time5, permutations = 9999, method = 'euclidian')
bact_perm_time5

#2) Ad Hoc Pairwise Analysis
pairwise_bac_time5 = pairwise.adonis2(bact_dist_time5~Treatment, data = sequencingkey_time5, permutations = 9999, method = 'euclidian')
pairwise_bac_time5

#3) Beta-dispersion: Distance from centroid of Group
#Calculating Beta-dispersion Distance from centroiod of Group)
bact_betadist_treat5 = betadisper(bact_dist_time5, group = sequencingkey_time5$Treatment)
distances_microbe_treat5 = data.frame(bact_betadist_treat5[['distances']])
distances_microbe_treat5 = cbind(distances_microbe_treat5, group = sequencingkey_time5$Treatment)
colnames(distances_microbe_treat5) = c('Distance_to_Centroid', 'Treatment')
#head(distances_microbe_treat5)

#Permutation test for homogeneity of Multivariate dispesions
test_bact_betadist_treatment5 = permutest(bact_betadist_treat5)
test_bact_betadist_treatment5

#Graphing Beta dispersions
B_Treat5 <- ggboxplot(data = distances_microbe_treat5, x = "Treatment", y = "Distance_to_Centroid" ) +
theme_classic() +
geom_point()
B_Treat5
```

Ad Hoc Pairwise analysis dataframe

```{r}
#Creating dataframe to hold pairwise comparisons
ad_hoc_pairwise <- data.frame(matrix(ncol =11))

#Naming columns of Dataframe
colnames(ad_hoc_pairwise) <-c("Treatment", "Week1_R2", "Week1_P","Week2_R2", "Week2_P", "Week3_R2", "Week3_P", "Week4_R2", "Week4_P", "Week5_R2", "Week5_P")

#Addng pairwise values
ad_hoc_pairwise[1,] <- c("SHAMP_vs_OVXP", pairwise_bac_time1$SHAMP_vs_OVXP$R2[1], pairwise_bac_time1$SHAMP_vs_OVXP$`Pr(>F)`[1], 
                         pairwise_bac_time2$SHAMP_vs_OVXP$R2[1], pairwise_bac_time2$SHAMP_vs_OVXP$`Pr(>F)`[1], 
                         pairwise_bac_time3$SHAMP_vs_OVXP$R2[1], pairwise_bac_time3$SHAMP_vs_OVXP$`Pr(>F)`[1],
                         pairwise_bac_time4$SHAMP_vs_OVXP$R2[1], pairwise_bac_time4$SHAMP_vs_OVXP$`Pr(>F)`[1],
                         pairwise_bac_time5$SHAMP_vs_OVXP$R2[1], pairwise_bac_time5$SHAMP_vs_OVXP$`Pr(>F)`[1])

ad_hoc_pairwise[2,] <- c("SHAMP_vs_SHAMDHT", pairwise_bac_time1$SHAMP_vs_SHAMDHT$R2[1], pairwise_bac_time1$SHAMP_vs_SHAMDHT$`Pr(>F)`[1],
                         pairwise_bac_time2$SHAMP_vs_SHAMDHT$R2[1], pairwise_bac_time2$SHAMP_vs_SHAMDHT$`Pr(>F)`[1],
                         pairwise_bac_time3$SHAMP_vs_SHAMDHT$R2[1], pairwise_bac_time3$SHAMP_vs_SHAMDHT$`Pr(>F)`[1], 
                         pairwise_bac_time4$SHAMP_vs_SHAMDHT$R2[1], pairwise_bac_time4$SHAMP_vs_SHAMDHT$`Pr(>F)`[1], 
                         pairwise_bac_time5$SHAMP_vs_SHAMDHT$R2[1], pairwise_bac_time5$SHAMP_vs_SHAMDHT$`Pr(>F)`[1])

ad_hoc_pairwise[3,] <- c("SHAMP_vs_OVXDHT", pairwise_bac_time1$SHAMP_vs_OVXDHT$R2[1], pairwise_bac_time1$SHAMP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time2$SHAMP_vs_OVXDHT$R2[1], pairwise_bac_time2$SHAMP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time3$SHAMP_vs_OVXDHT$R2[1], pairwise_bac_time3$SHAMP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time4$SHAMP_vs_OVXDHT$R2[1], pairwise_bac_time4$SHAMP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time5$SHAMP_vs_OVXDHT$R2[1], pairwise_bac_time5$SHAMP_vs_OVXDHT$`Pr(>F)`[1])

ad_hoc_pairwise[4,] <- c("OVXP_vs_SHAMDHT", pairwise_bac_time1$OVXP_vs_SHAMDHT$R2[1], pairwise_bac_time1$OVXP_vs_SHAMDHT$`Pr(>F)`[1],
                         pairwise_bac_time2$OVXP_vs_SHAMDHT$R2[1], pairwise_bac_time2$OVXP_vs_SHAMDHT$`Pr(>F)`[1],
                         pairwise_bac_time3$OVXP_vs_SHAMDHT$R2[1], pairwise_bac_time3$OVXP_vs_SHAMDHT$`Pr(>F)`[1],
                         pairwise_bac_time4$OVXP_vs_SHAMDHT$R2[1], pairwise_bac_time4$OVXP_vs_SHAMDHT$`Pr(>F)`[1],
                         pairwise_bac_time5$OVXP_vs_SHAMDHT$R2[1], pairwise_bac_time5$OVXP_vs_SHAMDHT$`Pr(>F)`[1])

ad_hoc_pairwise[5,] <- c("OVXP_vs_OVXDHT", pairwise_bac_time1$OVXP_vs_OVXDHT$R2[1], pairwise_bac_time1$OVXP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time2$OVXP_vs_OVXDHT$R2[1], pairwise_bac_time2$OVXP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time3$OVXP_vs_OVXDHT$R2[1], pairwise_bac_time3$OVXP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time4$OVXP_vs_OVXDHT$R2[1], pairwise_bac_time4$OVXP_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time5$OVXP_vs_OVXDHT$R2[1], pairwise_bac_time5$OVXP_vs_OVXDHT$`Pr(>F)`[1])

ad_hoc_pairwise[6,] <- c("SHAMDHT_vs_OVXDHT", pairwise_bac_time1$SHAMDHT_vs_OVXDHT$R2[1], pairwise_bac_time1$SHAMDHT_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time2$SHAMDHT_vs_OVXDHT$R2[1], pairwise_bac_time2$SHAMDHT_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time3$SHAMDHT_vs_OVXDHT$R2[1], pairwise_bac_time3$SHAMDHT_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time4$SHAMDHT_vs_OVXDHT$R2[1], pairwise_bac_time4$SHAMDHT_vs_OVXDHT$`Pr(>F)`[1],
                         pairwise_bac_time5$SHAMDHT_vs_OVXDHT$R2[1], pairwise_bac_time5$SHAMDHT_vs_OVXDHT$`Pr(>F)`[1])

ad_hoc_pairwise %>% write_csv(paste(lubridate::today(), 'Ad_hoc_table.csv', sep = "_"))
```

Taxa Clean Up

```{r}
#Subsetting only Important bacteria from Random forest
bacteria_new1 = subset(bacteria_new, select = c(names_rf))

#Merging Sequencing key for Identification
bacteria_new1 = merge(bacteria_new1, new_sequecingkey1, by=0)
bacteria_new1

#Removing Unnecessary level left in DF
bacteria_new1$Actual.Contents = droplevels(bacteria_new1$Actual.Contents)
bacteria_new1$Time = droplevels(bacteria_new1$Time)
bacteria_new1$Treatment = droplevels(bacteria_new1$Treatment)

#Setting Bacteria_new as genus for use in diff taxa analysis
genus_data <- bacteria_new1
genus_data = data.frame(genus_data, row.names = 1)


genus_data
```

Non-parametric taxa testing

```{r}
#Ranking the Data:
ranked_genus = genus_data %>%
  mutate(across(c(X9d607ec0281187001a9d6e5e22f9561d:f4496608bb586eb4cfd0969841385e32), rank))

#Loop to iterate over taxa for Linear model
n_features <- 50
H_null <- 'Time'
H_alt <- 'Treatment*Time'
rand <- 'Actual.Contents'

simple_lm <- data.frame(feature = character(),
                      Time_p = double(),
                      Treatment_p = logical(),
                      Inter_time_treatment = logical(),
                      OVXDHT_OVXP_P = double(),     # you will need a column for each 
                      OVXDHT_SHAMDHT_P = double(),   # categorical variable for both the 
                      OVXDHT_SHAMP_P = double(),   # here, there are 3 levels within the
                      OVXP_SHAMDHT_P = double(),        # you will need to change it below as
                      OVXP_SHAMP_P = double(),
                      SHAMDHT_SHAMP_P = double(),
                      stringsAsFactors = FALSE)

for (i in 1:n_features){
  
  feature = colnames(ranked_genus[i])
  fixed.alt <- formula(paste(feature,'~',H_alt))
  rando.form <- formula(paste('~1|', rand, sep=''))
  
  tryCatch(
        # TRY BLOCK
        { 
            # fitting the linear mixed effects model
            # alt model
            {alt_model <- lme(data = ranked_genus,
                      fixed = fixed.alt,
                      random = rando.form,
                      na.action=na.omit,
                      method='ML')}
            
        },
        # ERROR BLOCK
        error=function(e) {
            alt_model <- NULL
       })
 
        
  if(!is.null(alt_model)){
        #Type I ANOVA: Sequential sum of squares (two main factors (time and treatment) are tested in particular order, this type of squares will give different results for unbalanced data)
        comparison <- anova(alt_model)
        Time_p = comparison$`p-value`[2] #Time
        Treatment_p = comparison$`p-value`[3] #Treatment
        Inter_time_treatment=comparison$`p-value`[4] #Time:treatment interaction
        
        # perform multiple comparisons
        g <- emmeans(alt_model, pairwise ~ Treatment) 
        s <- summary(g)
        OVXDHT_OVXP_P = s$contrasts$p.value[1]
        OVXDHT_SHAMDHT_P = s$contrasts$p.value[2]
        OVXDHT_SHAMP_P = s$contrasts$p.value[3]
        OVXP_SHAMDHT_P = s$contrasts$p.value[4]
        OVXP_SHAMP_P = s$contrasts$p.value[5]
        SHAMDHT_SHAMP_P = s$contrasts$p.value[6]

        
  }
  df2 <- data.frame(feature, Time_p, Treatment_p, Inter_time_treatment, OVXDHT_OVXP_P, OVXDHT_SHAMDHT_P,OVXDHT_SHAMP_P, OVXP_SHAMDHT_P, OVXP_SHAMP_P, SHAMDHT_SHAMP_P)           # *** add columns here if needed
    rownames(df2) <- NULL
    names(df2) <- names(simple_lm)
    simple_lm = bind_rows(simple_lm, df2)
}

#adding X to feature name because R automatically adds it if number is starting
simple_lm$feature = gsub("X", "", as.character(simple_lm$feature))

#Merging feature table
simple_lm = merge(simple_lm, taxa, by="feature")
simple_lm = simple_lm %>% relocate(Taxon, .before = Time_p)
simple_lm = subset(simple_lm, select = -Confidence)


#FDR correction based on number of runs
simple_lm = simple_lm %>%
  mutate(`Time_p` = p.adjust(`Time_p`, method = 'fdr', n=50))%>%
  mutate(`Treatment_p` = p.adjust(`Treatment_p`, method = 'fdr', n=50))%>%
  mutate(`Inter_time_treatment` = p.adjust(`Inter_time_treatment`, method = 'fdr', n=50))

#writing to CSV file with data
simple_lm %>% write_csv(paste(lubridate::today(), 'HandzelTable3.csv', sep = "_"))

```

DID NOT USE CODE BELOW DUE TO NONNORMAL DATA!!

Differential Taxa Analysis for Normally Distributed Data

```{r}
n_features <- 50
H_null <- 'Time'
H_alt <- 'Treatment + Time'
rand <- 'Actual.Contents'

results <- data.frame(feature = character(),
                      main_p_value = double(),
                      Lowest_AIC = logical(),
                      Sig_to_H0 = logical(),
                      main_Slope = double(),
                      Normal_Dist = logical(),
                      Equal_var = logical(),
                      OVXDHT_OVXP_P = double(),     # you will need a column for each 
                      OVXDHT_SHAMDHT_P = double(),   # categorical variable for both the 
                      OVXDHT_SHAMP_P = double(),   # here, there are 3 levels within the
                      OVXP_SHAMDHT_P = double(),        # you will need to change it below as
                      OVXP_SHAMP_P = double(),
                      SHAMDHT_SHAMP_P = double(),
                      stringsAsFactors = FALSE)

for (i in 1:n_features){
  
  feature = colnames(genus_data[i])
  #print(feature)
  fixed.null <- formula(paste(feature,'~',H_null))
  fixed.alt <- formula(paste(feature,'~',H_alt))
  rando.form <- formula(paste('~1|', rand, sep=''))
  
  tryCatch(
        # TRY BLOCK
        { 
            # fitting the linear mixed effects model
            # null model
            {null_model <- lme(data = genus_data,
                      fixed = fixed.null,
                      random = rando.form,
                      na.action=na.omit,
                      method='ML')}
            
            {alt_model <- lme(data = genus_data,
                     fixed = fixed.alt,
                     random = rando.form,
                     na.action=na.omit,
                     method='ML')}
        },
        # ERROR BLOCK
        error=function(e) {
            null_model <- NULL
            alt_model <- NULL
       })
 
        
  if(!is.null(alt_model)){
        # compare LMMs and get T/F lowest AIC 
        comparison <- anova(null_model, alt_model)
        #Checks if alt model is a lower AIC means its a better model
        Lowest_AIC <- (which.min(comparison$AIC) == 2) 
        #Checks p-value of alt model
        null_comp <- (comparison[2,9] <= 0.05)
        
        # get p-values from ANOVA type 2
        #Split total variation of dependent vaiables into different sources of variation
        #Allows us to find out whether independent variable have significant effect on dependent variable
        test <- car::Anova(alt_model, type = 2)
        main_p_val <- test[1,3]
        
        # get coefficients
        main_slope <- coef(alt_model)[1,2]
        
        # normality assessment
        normal <- (shapiro.test(resid(alt_model))$p.value > 0.05)
    
        # homogeneity of variance, H0: equal
        # https://ademos.people.uic.edu/Chapter18.html#62_assumption_2_homogeneity_of_variance
        genus_data$sq.resid <- abs(resid(alt_model))^2
        levene <- lm(sq.resid ~ Treatment, data = genus_data)
        #Getting p-Value: IF greater than 0.05 variance of the residuals is equal (assumption of homoscedasticity is met)
        equal_var <- (anova(levene)[1,5] > 0.05)
        
        # if main p-value is <= 0.05, perform multiple comparisons
        if(main_p_val<=0.05 & normal==TRUE & equal_var==TRUE){
            g <- emmeans(alt_model, pairwise ~ Treatment) 
            s <- summary(g)
            OVXDHT_OVXP_P = s$contrasts$p.value[1]
            OVXDHT_SHAMDHT_P = s$contrasts$p.value[2]
            OVXDHT_SHAMP_P = s$contrasts$p.value[3]
            OVXP_SHAMDHT_P = s$contrasts$p.value[4]
            OVXP_SHAMP_P = s$contrasts$p.value[5]
            SHAMDHT_SHAMP_P = s$contrasts$p.value[6]
        } else {
            OVXDHT_OVXP_P = NA
            OVXDHT_SHAMDHT_P = NA
            OVXDHT_SHAMP_P = NA
            OVXP_SHAMDHT_P = NA
            OVXP_SHAMP_P = NA
            SHAMDHT_SHAMP_P = NA
        }
        
  }
  df <- data.frame(feature, main_p_val, Lowest_AIC, null_comp, main_slope, normal, equal_var, OVXDHT_OVXP_P, OVXDHT_SHAMDHT_P,OVXDHT_SHAMP_P, OVXP_SHAMDHT_P, OVXP_SHAMP_P, SHAMDHT_SHAMP_P)           # *** add columns here if needed
    rownames(df) <- NULL
    names(df) <- names(results)
    results = bind_rows(results, df)
    
}


```

DID NOT USE Code Below

Taxa analysis if you transformed data and it was normal!

```{r}
n_features <- 50
H_null <- 'Time'
H_alt <- 'Treatment + Time'
rand <- 'Actual.Contents'

xformed <- data.frame(feature = character(),
                      main_p_value = double(),
                      Lowest_AIC = logical(),
                      Sig_to_H0 = logical(),
                      main_Slope = double(),
                      Normal_Dist = logical(),
                      Equal_var = logical(),
                      OVXDHT_OVXP_P = double(),     # you will need a column for each 
                      OVXDHT_SHAMDHT_P = double(),   # categorical variable for both the 
                      OVXDHT_SHAMP_P = double(),   # here, there are 3 levels within the
                      OVXP_SHAMDHT_P = double(),        # you will need to change it below as
                      OVXP_SHAMP_P = double(),
                      SHAMDHT_SHAMP_P = double(),
                      stringsAsFactors = FALSE)

for (i in 1:n_features){
  
  feature = colnames(ranked_genus[i])
  #print(feature)
  fixed.null <- formula(paste(feature,'~',H_null))
  fixed.alt <- formula(paste(feature,'~',H_alt))
  rando.form <- formula(paste('~1|', rand, sep=''))
  
  tryCatch(
        # TRY BLOCK
        { 
            # fitting the linear mixed effects model
            # null model
            {null_model <- lme(data = ranked_genus,
                      fixed = fixed.null,
                      random = rando.form,
                      na.action=na.omit,
                      method='ML')}
            
            {alt_model <- lme(data = ranked_genus,
                     fixed = fixed.alt,
                     random = rando.form,
                     na.action=na.omit,
                     method='ML')}
        },
        # ERROR BLOCK
        error=function(e) {
            null_model <- NULL
            alt_model <- NULL
       })
 
        
  if(!is.null(alt_model)){
        # compare LMMs and get T/F lowest AIC 
        comparison <- anova(null_model, alt_model)
        #Checks if alt model is a lower AIC means its a better model
        Lowest_AIC <- (which.min(comparison$AIC) == 2) 
        #Checks p-value of alt model
        null_comp <- (comparison[2,9] <= 0.05)
        
        # get p-values from ANOVA type 2
        #Split total variation of dependent vaiables into different sources of variation
        #Allows us to find out whether independent variable have significant effect on dependent variable
        test <- car::Anova(alt_model, type = 2)
        main_p_val <- test[1,3]
        
        # get coefficients
        main_slope <- coef(alt_model)[1,2]
        
        # normality assessment
        normal <- (shapiro.test(resid(alt_model))$p.value > 0.05)
    
        # homogeneity of variance, H0: equal
        # https://ademos.people.uic.edu/Chapter18.html#62_assumption_2_homogeneity_of_variance
        ranked_genus$sq.resid <- abs(resid(alt_model))^2
        levene <- lm(sq.resid ~ Treatment, data = ranked_genus)
        #Getting p-Value: IF greater than 0.05 variance of the residuals is equal (assumption of homoscedasticity is met)
        equal_var <- (anova(levene)[1,5] > 0.05)
        
        # if main p-value is <= 0.05, perform multiple comparisons
        g <- emmeans(alt_model, pairwise ~ Treatment) 
        s <- summary(g)
        OVXDHT_OVXP_P = s$contrasts$p.value[1]
        OVXDHT_SHAMDHT_P = s$contrasts$p.value[2]
        OVXDHT_SHAMP_P = s$contrasts$p.value[3]
        OVXP_SHAMDHT_P = s$contrasts$p.value[4]
        OVXP_SHAMP_P = s$contrasts$p.value[5]
        SHAMDHT_SHAMP_P = s$contrasts$p.value[6]

        
  }
  df1 <- data.frame(feature, main_p_val, Lowest_AIC, null_comp, main_slope, normal, equal_var, OVXDHT_OVXP_P, OVXDHT_SHAMDHT_P,OVXDHT_SHAMP_P, OVXP_SHAMDHT_P, OVXP_SHAMP_P, SHAMDHT_SHAMP_P)           # *** add columns here if needed
    rownames(df1) <- NULL
    names(df1) <- names(xformed)
    xformed = bind_rows(xformed, df1)
}
```

Creating Table with taxa, FDR corrected P-value, and LMM p-values

```{r}
xformed$feature = gsub("X", "", as.character(xformed$feature))
xformed = xformed %>%
  mutate(main_p_value = p.adjust(main_p_value, method = 'fdr'))%>%
  filter(main_p_value <= 0.05)

xformed = merge(xformed, taxa, by="feature")
xformed = xformed %>% relocate(Taxon, .before = main_p_value)
xformed = subset(xformed, select = -Confidence)
xformed %>% write_csv(paste(lubridate::today(), 'LMM_taxa_table.csv', sep = "_"))
```


